### PR TITLE
Add search and JQL functions to Jira bot

### PR DESCRIPTION
**Pull request by pemontto**
This PR adds functionality for searching issue summaries with `search` and full blown JQL search with `jql`.

It also supports the user defining their own HTTP scheme in the `domain`, this is useful for circumstances where Jira is served behind an authenticating proxy. This allows you to connect to a secondary listener that might not be HTTPS. An optional configuration parameter was also added `display_url` so the URL for the output links can also be controlled.

E.g. Users access Jira through a proxy at `https://jira.mydomain.com`. The actual Jira host listens on `http://jira1.mydomain.com:8081`. You can now configure this using
```
[jira]
username = test
password = test_password
domain = http://jira1.mydomain.com:8081
display_url = https://jira.mydomain.com
```

### CODE REVIEW

File: zulip_bots/zulip_bots/bots/jira/jira.conf [hunk 1]
```diff
@@ -2,3 +2,4 @@
 username = <your Jira email address>
 password = <your Jira password>
 domain = <your Jira domain>
+# domain_url = [optional] <your front facing Jira URL to use in links>
```
Comment (aero31aero) [inline zulip_bots/zulip_bots/bots/jira/jira.conf:5]: Missing new line at the and of this file.

File: zulip_bots/zulip_bots/bots/jira/jira.py [hunk 2]
```diff
@@ -127,7 +163,40 @@ def initialize(self, bot_handler: Any) -> None:
             raise KeyError('No `domain` was specified')
 
         self.auth = make_jira_auth(username, password)
-        self.domain_with_protocol = 'https://' + domain
+
+        # Allow users to override the HTTP scheme
```
Comment (aero31aero) [inline zulip_bots/zulip_bots/bots/jira/jira.py:167]: It would be better to split this into its own commit. I'm not sure if this a good change; we should try to encourage using https and not provide an easy way to override that.

File: zulip_bots/zulip_bots/bots/jira/jira.py [hunk 3]
```diff
@@ -127,7 +163,40 @@ def initialize(self, bot_handler: Any) -> None:
...
+
+        # Use the front facing URL in output
+        self.display_url = config.get('display_url')
+        if not self.display_url:
+            self.display_url = self.domain_with_protocol
+
+    def jql_search(self, jql_query:str) -> str:
+        UNKNOWN_VAL = '*unknown*'
+        jira_response = requests.get(
+            f'{self.domain_with_protocol}/rest/api/2/search?jql={jql_query}&fields=key,summary,status',
```
Comment (aero31aero) [inline zulip_bots/zulip_bots/bots/jira/jira.py:181]: Unfortunately, we cannot use f-strings yet. We have to support python 3.5 as well, thus, we should use `"{}".format()` for now.

File: zulip_bots/zulip_bots/bots/jira/jira.py [hunk 4]
```diff
@@ -127,7 +163,40 @@ def initialize(self, bot_handler: Any) -> None:
...
+        ).json()
+
+        url = self.display_url + '/browse/'
+        errors = jira_response.get('errorMessages', [])
+        results = jira_response.get('total', 0)
+
+        if errors:
+            response = 'Oh no! Jira raised an error:\n > ' + ', '.join(errors)
+        else:
+            response = f'*Found {results} results*\n\n'
```
Comment (aero31aero) [inline zulip_bots/zulip_bots/bots/jira/jira.py:192]: Similarly here.

File: zulip_bots/zulip_bots/bots/jira/jira.py [hunk 5]
```diff
@@ -127,7 +163,40 @@ def initialize(self, bot_handler: Any) -> None:
...
-        self.domain_with_protocol = 'https://' + domain
+
+        # Allow users to override the HTTP scheme
+        if re.match(r'^https?://', domain, re.IGNORECASE):
+            self.domain_with_protocol = domain
+        else:
+            self.domain_with_protocol = 'https://' + domain
+
+        # Use the front facing URL in output
+        self.display_url = config.get('display_url')
```
Comment (aero31aero) [inline zulip_bots/zulip_bots/bots/jira/jira.py:174]: Similarly, it makes sense to have adding display_url as a separate commit.

File: zulip_bots/zulip_bots/bots/jira/jira.conf (hunk 1 above)
Comment (aero31aero) [reply zulip_bots/zulip_bots/bots/jira/jira.conf:5]: You probably meant to write `display_url` here.

**Comment by pemontto** [comment]
Thanks for the quick review, I was hesitant to spam the commit history. Will address those issues now.

File: zulip_bots/zulip_bots/bots/jira/jira.conf (hunk 1 above)
Comment (pemontto) [reply zulip_bots/zulip_bots/bots/jira/jira.conf]: Fixed

**Comment by timabbott** [comment]
Merged, after squashing that last documentation fixup commit, thanks @pemontto!
